File: backend/app/models/customer.py
import enum
import uuid

from sqlalchemy import ForeignKey, String, Text, Boolean, Enum
from sqlalchemy.dialects.postgresql import UUID, ARRAY
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import UUIDBase


class CustomerStatus(str, enum.Enum):
    ACTIVE = "active"
    BLOCKED = "blocked"
    MUTED = "muted"
    BLACKLISTED = "blacklisted"


class Customer(UUIDBase):
    __tablename__ = "customers"

    business_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("businesses.id"), index=True
    )
    wa_id: Mapped[str] = mapped_column(String(50), index=True)
    phone: Mapped[str] = mapped_column(String(20))
    name: Mapped[str | None] = mapped_column(String(150))
    email: Mapped[str | None] = mapped_column(String(255))
    notes: Mapped[str | None] = mapped_column(Text)

    # Customer control fields
    status: Mapped[CustomerStatus] = mapped_column(
        Enum(CustomerStatus), default=CustomerStatus.ACTIVE
    )
    ai_enabled: Mapped[bool] = mapped_column(Boolean, default=True)
    is_flagged: Mapped[bool] = mapped_column(Boolean, default=False)
    tags: Mapped[list[str] | None] = mapped_column(ARRAY(String(50)), nullable=True)
    block_reason: Mapped[str | None] = mapped_column(String(255))

    business: Mapped["Business"] = relationship(back_populates="customers")
    conversations: Mapped[list["Conversation"]] = relationship(back_populates="customer")
    orders: Mapped[list["Order"]] = relationship(back_populates="customer")

File: mobile/lib/models/customer.dart
class Customer {
  final String id;
  final String name;
  final String? phone;
  final String? email;
  final String status;
  final bool aiEnabled;
  final bool isFlagged;
  final List<String> tags;
  final int conversationCount;
  final int orderCount;
  final String createdAt;

  Customer({
    required this.id,
    required this.name,
    this.phone,
    this.email,
    this.status = 'active',
    this.aiEnabled = true,
    this.isFlagged = false,
    this.tags = const [],
    this.conversationCount = 0,
    this.orderCount = 0,
    required this.createdAt,
  });

  factory Customer.fromJson(Map<String, dynamic> json) => Customer(
        id: json['id'] as String,
        name: json['name'] as String? ?? 'Unknown',
        phone: json['phone'] as String?,
        email: json['email'] as String?,
        status: json['status'] as String? ?? 'active',
        aiEnabled: json['ai_enabled'] as bool? ?? true,
        isFlagged: json['is_flagged'] as bool? ?? false,
        tags: (json['tags'] as List<dynamic>?)
                ?.map((t) => t.toString())
                .toList() ??
            [],
        conversationCount: json['conversation_count'] as int? ?? 0,
        orderCount: json['order_count'] as int? ?? 0,
        createdAt: json['created_at'] as String,
      );
}
